react/logic/formState.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { useReducer } from 'react'
+
+const initialState: SubmittingState = {
+  loading: false,
+  userInputError: null,
+  serverError: null,
+  success: null,
+}
+function reducer(
+  state: SubmittingState,
+  action: SubmittingAction
+): SubmittingState {
+  switch (action.type) {
+    case 'START_SUBMITTING': {
+      return {
+        loading: true,
+        userInputError: null,
+        serverError: null,
+        success: null,
+      }
+    }
+    case 'SUCCESS_SUBMITTING': {
+      return {
+        loading: false,
+        userInputError: false,
+        serverError: null,
+        success: true,
+      }
+    }
+    case 'USER_INPUT_ERROR_SUBMITTING': {
+      return {
+        ...state,
+        loading: false,
+        userInputError: true,
+        success: false,
+      }
+    }
+    case 'SERVER_INTERNAL_ERROR_SUBMITTING': {
+      return {
+        ...state,
+        loading: false,
+        serverError: true,
+        success: false,
+      }
+    }
+    default:
+      return state
+  }
+}
+
+export type SubmittingState = {
+  loading: boolean
+  userInputError: boolean | null
+  serverError: boolean | null
+  success: boolean | null
+}
+
+export type SubmittingAction =
+  | { type: 'START_SUBMITTING' }
+  | { type: 'SUCCESS_SUBMITTING' }
+  | { type: 'USER_INPUT_ERROR_SUBMITTING' }
+  | { type: 'SERVER_INTERNAL_ERROR_SUBMITTING' }
+
```

Comment:
These names are sounding a bit weird to me. Some of them aren't actually verbs and it is not very clear what they are doing